### PR TITLE
Use strings concatenate to support large strings in CSV writer

### LINKED ISSUE #16137: [FEA] Add large strings support in CSV writer
**Is your feature request related to a problem? Please describe.**
The libcudf CSV writer throws a `std::overflow_error` when trying to file a dataframe or a chunk that exceeds 2.1B characters. 

**Describe the solution you'd like**
There are a few options:
1. fix the overflow error in the CSV writer implementation. This would be a libcudf change. Since the CSV writer is fairly simple and relies on libcudf strings API, there may be a straightforward solution.
2. Change the default chunksize to libcudf or cuDF from `std::numeric_limits<size_type>::max();` to something based on the dataframe size, e.g. `chunksize = len(df) // (df.memory_usage(deep=True).sum() / 500_000_000.)`

**Additional context**
As far as fixing the root cause, here is a quick repro:
```
    df = cudf.DataFrame({'text': ['a'] * 1000})
    df['text'] = df['text'].str.repeat(1000)
    data_1gb = cudf.concat([df] * 1000, ignore_index=True)
    df = cudf.concat([data_1gb] * 3, ignore_index=True)
    df.to_csv('/raid/tmp.csv')
```
which throws:
```
  File "csv.pyx", line 547, in cudf._lib.csv.write_csv
OverflowError: Writing CSV file with chunksize=3000000 failed. Consider providing a smaller chunksize argument.
```
but I couldn't immediately find where the `std::overflow_error` is getting thrown in `io/csv/writer_impl.cu`.

As far as chunksize, I did some analysis of chunksize, and using smaller chunks of ~100 MB shouldn't have a significant performance impact.
image

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Changes the CSV writer logic to use `cudf::strings::concatenate` instead of `cudf::strings::join_strings` when output size exceeds `join_strings` limit.

Closes #16137 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by GregoryKimball** [comment]
Thank you @davidwendt for jumping on this! Amazing work.

**Comment by davidwendt** [comment]
`CSV_WRITER_NVBENCH` results for reference
```
## [0] Quadro GV100

|  na_per_len  |  rows_per_chunk  |   Ref Time |   Cmp Time |       Diff |   %Diff |
|--------------|------------------|------------|------------|------------|---------|
|      0       |       2^8        |   10.079 s |   10.173 s |  93.966 ms |   0.93% |
|      16      |       2^8        |   10.197 s |   10.255 s |  57.880 ms |   0.57% |
|      0       |       2^10       |    3.930 s |    4.005 s |  74.697 ms |   1.90% |
|      16      |       2^10       |    3.917 s |    4.041 s | 123.625 ms |   3.16% |
|      0       |       2^12       |    2.028 s |    2.070 s |  42.833 ms |   2.11% |
|      16      |       2^12       |    1.928 s |    1.973 s |  44.713 ms |   2.32% |
|      0       |       2^14       |    1.448 s |    1.486 s |  37.474 ms |   2.59% |
|      16      |       2^14       |    1.643 s |    1.647 s |   3.840 ms |   0.23% |
|      0       |       2^16       |    1.492 s |    1.558 s |  65.690 ms |   4.40% |
|      16      |       2^16       |    1.520 s |    1.538 s |  18.826 ms |   1.24% |
|      0       |       2^18       |    1.928 s |    1.968 s |  39.460 ms |   2.05% |
|      16      |       2^18       |    1.643 s |    1.679 s |  35.551 ms |   2.16% |
|      0       |       2^20       |    1.389 s |    1.418 s |  28.738 ms |   2.07% |
|      16      |       2^20       |    1.406 s |    1.432 s |  26.335 ms |   1.87% |

```

**Comment by GregoryKimball** [comment]
Thank you @davidwendt, why not just use `concatenate` for both small strings and large strings? Is the performance impact significant between `join_strings` and `concatenate`?

**Comment by davidwendt** [comment]
> Thank you @davidwendt, why not just use `concatenate` for both small strings and large strings? Is the performance impact significant between `join_strings` and `concatenate`?

Yes, the `concatenate` showed up to 25% performance degradation in the benchmark.
Technically it would be best to use a smaller chunksize to keep each output chunk less than 2GB.

File: cpp/src/io/csv/writer_impl.cu [hunk 1]
```diff
@@ -491,7 +510,8 @@ void write_csv(data_sink* out_sink,
           str_table_view.column(0), options_narep, stream, rmm::mr::get_current_device_resource());
       }();
 
-      write_chunked(out_sink, str_concat_col->view(), options, stream, mr);
+      write_chunked(
+        out_sink, str_concat_col->view(), options, stream, rmm::mr::get_current_device_resource());
```
Comment (ttnghia) [inline cpp/src/io/csv/writer_impl.cu:514]: It makes sense to use the default device resource for writing data, but then why do we need to have an `mr` parameter above?

Comment (davidwendt) [reply cpp/src/io/csv/writer_impl.cu:514]: Yes, I noticed that too. This function returns a void so there should actually be no need to pass a `mr` at all perhaps. That seemed like a bigger change out of scope of this PR so I was going to explore it in a different issue/PR. What do you think?

Comment (ttnghia) [reply cpp/src/io/csv/writer_impl.cu:514]: Yes that sounds good. Thanks.

**Review by ttnghia — APPROVED** [review]

**Review by PointKernel — APPROVED** [review]

**Pull request merged**